net/sqlwriter.c

```diff
@@ -418,14 +399,9 @@ struct sqlwriter *sqlwriter_new(struct sqlwriter_arg *arg)
     writer->wr_base = event_base_new_with_config(cfg);
     event_config_free(cfg);
 
-    int flush_flags = EV_WRITE | EV_TIMEOUT | EV_PERSIST;
-    writer->flush_ev = event_new(writer->wr_base, arg->fd, flush_flags, sql_flush_cb, writer);
-
-    int hb_flags = EV_TIMEOUT | EV_PERSIST;
-    writer->heartbeat_ev = event_new(appsock_timer_base, -1, hb_flags, sql_heartbeat_cb, writer);
-
-    int trickle_flags = EV_WRITE | EV_PERSIST;
-    writer->heartbeat_trickle_ev = event_new(appsock_timer_base, arg->fd, trickle_flags, sql_trickle_cb, writer);
+    writer->flush_ev = event_new(writer->wr_base, arg->fd, EV_WRITE, sql_flush_cb, writer);
```

Comment:
Wouldn't it be better to keep this persistent - so we don't have to `event_add` in a loop in `sql_flush_int`

Reply:
The flush event doesn't gets fired repeatedly. Generally it triggers only on the last row. So it makes sense to make it not persistent such that it'll be automatically removed from queue once triggered.  If the event is PERSIST, we'll have to call event_del to remove it from the event base, which is a lot more expensive.

You're right that we may have to `event_add` in the while loop if the event isn't PERSIST. However the chances of  `writer->bad` are relatively low, so in most cases we only `event_add` once.

Reply:
_However the chances of writer->bad_
I wasn't thinking about `writer->bad`. Won't this require scheduling flush a few times to write the entire buffered response. I was thinking one (event_add + event_del) vs (several event_add). What does a table scan look like? Or even a typical request (I don't know what that is.. 10k rows or 1K each?) Put a counter and test?

Reply:
We're doing [a while loop](https://github.com/bloomberg/comdb2/blob/master/net/sqlwriter.c#L156) in `sql_flush_cb()` to write out the evbuffer. So unless there's a slow reader, the `flush_ev` event gets scheduled once per flush (Though `evbuffer_write()` may be called multiple times during each run).

However I'm removing this while loop to make this logic more non-blocking-like: if insufficient bytes are written, exit  `sql_flush_cb()`; the callback will be picked up again on next EV_READ event. I've also changed flush_ev back to EV_PERSIST. My benchmark shows that one (event_add + event_del) and several event_add's here perform almost identically.

Reply:
I don't think that while loop is sufficient. That only fills up sending tcp buffer and more than that amount can be easily outstanding - even if reader is fast.

Reply:
The while loop exists to paper over limited number of iovecs that writev will process. So with lots of small rows, tcp buffer is not full, but one writev doesn't pick up all the iovecs. At least, that was what I observed when I added that loop. I don't think removing it achieves anything - Idea was, we are flushing in our own dedicated thread, why go through another epoll when tcp buff isn't full.

Reply:
Which is also why trickle tries to make the outstanding data contiguous.. But I'm less sure about how useful this is.

Reply:
Thank you! I wasn't aware of the writev limit. I put the while loop back.